test/charms/test_main/src/charm.py

```diff
@@ -138,6 +144,19 @@ def on_foo_bar_action(self, event):
         self._state['observed_event_types'].append(type(event))
         self._write_state()
 
+    def on_collect_metrics(self, event):
+        self._state['on_collect_metrics'].append(type(event))
+        self._state['observed_event_types'].append(type(event))
+        event.add_metrics({'foo': 'bar'}, {'dead': ' beef '})
+        self._write_state()
+
+    def on_ha_relation_departed(self, event):
+        indicator_file = pathlib.Path(self._charm_config['INDICATOR_FILE'])
+        indicator_file.touch()
+        with open(self._state_file, 'w+') as state_fd:
+            fcntl.flock(state_fd, fcntl.LOCK_EX)
+            fcntl.flock(state_fd, fcntl.LOCK_UN)
```

Comment:
How's that related to the metric collection?

Reply:
https://github.com/canonical/operator/pull/140#discussion_r379516118